EpistasisLab/tpot
# Implementation of feature extraction

## What does this PR do?

This PR implements feature extraction in TPOT, which enables it to natively support image and text data while constructing its pipelines. It does this by modifying the GP logic to check for the operator types throughout the process and then considering the input to the tree when selecting the final operator.


## Where should the reviewer start?

`tpot/builtins/feature_extractors.py` has the base code for the implementation of the feature extractors. 

`tpot/base.py` has many extensive changes to support the modified genetic programming logic as well as checks for input type to allow for different input checking depending on the input type (allowing an array of strings for text input; allowing n-dimensional inputs for images).

`tests/feature_extract_tests.py` and `tests/tpot_complex_input_tests.py` includes sets of development tests for checking that feature extraction works.

Many documentation files in `docs_sources` have been changed to include all documentation needed for an end-user to use feature extraction as well 


## How should this PR be tested?

This was developed in-house and has been tested on all of TPOT's base tests as well as the new tests developed for feature extraction. We are merging this into a separate branch to ensure compatibility with future versions of TPOT, but this can operate as a drop-in replacement for TPOT 0.11.7.

For others who are interested, you can run the nosetests yourself to ensure that this passes all relevant tests.


## Any background context you want to provide?

This was developed by a graduate student at Penn while in the Epistasis Lab.


## What are the relevant issues?

N/A


## Questions:

- Do the docs need to be updated? - Already updated
- Does this PR add new (Python) dependencies? - Can include torch, but will function properly without torch if needed. All other extractors do not require new dependencies.


@JDRomano2
